play-json/shared/src/main/scala/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -519,7 +519,14 @@ import scala.reflect.macros.blackbox
         c.abort(c.enclosingPosition, s"Sealed trait ${atpe} is not supported: no known subclasses")
       }
 
-      val typeNaming = (_: Type).typeSymbol.fullName
+      def typeNaming(t: Type): String = {
+        /** Extract "name" from @Discriminator("name") */
```

Comment:
I would rather have a `subtypeFqdn: String => discriminatorValue: String` as a naming strategy that can be specified thanks to `JsonConfiguration`.

It allows to manage differents namings for a dame sealed family according different use cases.

Also to have all the settings at once/in one place with the `JsonConfiguration`.

> For me it can be considered later whether to automatically generate a default naming strategy for discriminator values of types according annotation.

Reply:
I did consider this. It's useful if the discriminators can be derived from the class name, e.g. `fully.qualified.CaseClassName` => `"case-class-name"`. For my use-case, I prefer the annotation over a nasty pattern match with no exhaustiveness/correctness checking.

But I do agree with you. For a consistent API we should support the same options for class names as we have for the class member names. What do you think about this proposal?

1. Support `JsonNaming` for the class name:
```scala
implicit val c = JsonConfiguration(
  naming = JsonNaming.SnakeCase,
  classNaming = JsonNaming.SnakeCase,
  discriminator = "_disc",
)
trait Base
case class MyCaseClass(myClassMember: Int) extends Base
// { "_disc": "my_case_class", "my_class_member": 42 }
```

2. Support compile-time annotations for both class names and member names, with the annotation taking priority over other naming configuration:
```scala
trait Base
@Discriminator("custom") case class MyCaseClass(@Name("mbr") myClassMember: Int) extends Base
// { "_type": "custom", "mbr": 42 }
```

Reply:
I'm more for option 1, with everything being configurable homogeneously in `JsonConfiguration` with `JsonNaming` strategies.

Annotation can be discussed worked with a distinct PR.